src/sidebar/components/tag-editor.js

```diff
@@ -1,35 +1,202 @@
+      const suggestionsSet = [];
+      suggestions.forEach(suggestion => {
+        if (tagList.indexOf(suggestion) < 0) {
+          suggestionsSet.push(suggestion);
+        }
       });
+      return suggestionsSet;
+    };
+
+    if (e.key && e.key !== 'Enter') {
+      // A user typed a letter and we need to filter the results
+      // to update the <dataset> suggestions
+      const suggestions = removeDuplicates(
+        tagsService.filter(inputEl.current.value)
+      );
+      setSuggestions(suggestions);
+    } else if (e.type === 'keyup') {
+      // Assume the user selected an option from <datalist>
+      // either by clicking on it with the mouse or pressing
+      // the 'Enter' key.
+      addTag();
     }
   };
+
+  const buildSuggestionsList = () => {
+    const suggestionElements = suggestions.map((suggestion, index) => (
+      <option key={index} value={suggestion} />
+    ));
+
+    // Only return a datalist if there is something to render in the list. This
+    // avoid a bug in Firefox where the datalist rendered does not correspond to the
+    // actual list of <option>s in the dom. Wiping out the entire <datalist> tag seems
+    // to force Firefox's hand.
+    if (suggestionElements.length > 0) {
+      return (
+        <datalist
+          id={`tag-editor-datalist-${annotation.id}`}
+          className="tag-editor__suggestions"
+          aria-label="Annotation suggestions"
+        >
+          {suggestionElements}
+        </datalist>
+      );
+    }
+    return null;
+  };
+
+  return (
+    <section className="tag-editor">
+      <ul className="tag-editor__tag-list" aria-label="Annotation tags">
+        {tagList.map(tag => {
+          return (
+            <li
+              key={`${tag}`}
+              className="tag-editor__tag-item"
+              aria-label={`Tag: ${tag}`}
+            >
+              <span className="tag-editor__edit">{tag}</span>
+              <button
+                onClick={() => {
+                  removeTag(tag);
+                }}
+                title={`Remove Tag: ${tag}`}
+                className="tag-editor__delete"
+              >
+                <SvgIcon name="cancel" />
+              </button>
+            </li>
+          );
+        })}
+      </ul>
+      <input
+        list={`tag-editor-datalist-${annotation.id}`}
+        onBlur={handleKeyUp}
+        onKeyUp={handleKeyUp}
+        onKeyPress={handleKeyPress}
+        ref={inputEl}
+        placeholder="Add tags..."
+        className="tag-editor__input"
+        type="text"
+      />
+      {buildSuggestionsList()}
+    </section>
+  );
 }
 
-module.exports = {
-  controller: TagEditorController,
-  controllerAs: 'vm',
-  bindings: {
-    tags: '<',
-    onEditTags: '&',
-  },
-  template: require('../templates/tag-editor.html'),
+/**
+ * @typedef Tag
+ * @param tag {string} - The tag text
+ */
+
+TagEditor.propTypes = {
+  /* Annotation that owns the tags. */
+  annotation: propTypes.object,
```

Comment:
None of these are marked as `isRequired`. How does the component behave if any are missing? Seems like they might all need to be required?